### PR TITLE
Get rid of hardcoded float in the geometry algorithms

### PR DESCRIPTION
**Pull request by aprokop**
Also replace std::sqrt by Kokkos::sqrt to allow for half_t and bhalf_t calculations in the future.

### CODE REVIEW

File: src/geometry/ArborX_DetailsAlgorithms.hpp [hunk 1]
```diff
@@ -195,24 +195,25 @@ struct isValid<SphereTag, Sphere>
 template <typename Point1, typename Point2>
 struct distance<PointTag, PointTag, Point1, Point2>
 {
-  KOKKOS_FUNCTION static float apply(Point1 const &a, Point2 const &b)
+  KOKKOS_FUNCTION static auto apply(Point1 const &a, Point2 const &b)
   {
     constexpr int DIM = GeometryTraits::dimension_v<Point1>;
-    float distance_squared = 0.0;
+    using Coordinate = typename GeometryTraits::coordinate_type<Point1>::type;
+    Coordinate distance_squared = 0;
```
Comment (dalg24) [inline src/geometry/ArborX_DetailsAlgorithms.hpp:202]: Presumably if you want to use that coordinate type for the calculation you should cast b[d] and a[d] below.

We don't seem to enforce that the two points types are the same.  Just using the coordinate type of the first point does not feel right.

Comment (aprokop) [reply src/geometry/ArborX_DetailsAlgorithms.hpp:202]: Guess we can use the type of `a[d] - b[d]` to allow implicit conversions.

**Review by masterleinad — APPROVED** [review]
Looks reasonable to me.

**Review by dalg24 — APPROVED** [review]
Make sure you squash merge

**Comment by aprokop** [comment]
`Clang` build did not start. All others pass.

**Pull request merged**